lib/src/tag_detection.dart

```diff
@@ -983,23 +919,90 @@ class Tagger {
     }
   }
 
+  /// Decides if a package is null-safe.
+  ///
+  /// A package is marked null-safety compliant if:
+  ///
+  /// - The package and all its transitive dependencies have opted-in by
+  ///   specifying a lower dart sdk bound >= 2.12.
+  ///
+  /// - No libraries in the import closure of _topLibraries opt out of
+  ///   null-safety. (For each runtime in [Runtime._recognizedRuntimes]).
   void nullSafetyTags(List<String> tags, List<Explanation> explanations) {
+    const _nullSafeTag = 'is:null-safe';
+
     try {
-      final nullSafety =
-          _NullSafetyViolationFinder(_packageGraph, _pubspecCache, _session);
-      final nullSafetyResult = nullSafety.findNullSafetyViolation(packageName);
-      if (nullSafetyResult != null) {
-        explanations.add(nullSafetyResult);
+      var foundIssues = false;
+
+      final sdkConstraintFinder = _PathFinder<String>(_packageGraph, (
+        String packageDir,
+      ) {
+        final pubspec = _pubspecCache.pubspecOfPackage(packageDir);
+
+        return pubspec.sdkConstraintStatus.hasOptedIntoNullSafety
+            ? null
+            : (path) => Explanation(
+                  'Package is not null safe',
+                  'Because:\n${_PackageGraph.formatPath(path)} '
+                      'that doesn\'t opt in to null safety',
+                  tag: _nullSafeTag,
+                );
+      });
+
+      final sdkConstraintResult =
+          sdkConstraintFinder.findViolation(packageName);
+      if (sdkConstraintResult != null) {
+        explanations.add(sdkConstraintResult);
+        foundIssues = true;
       } else {
-        return tags.add(_NullSafetyViolationFinder.nullSafeTag);
+        for (final runtime in Runtime._recognizedRuntimes) {
+          final optOutViolationFinder = _PathFinder<Uri>(
+            LibraryGraph(_session, runtime.declaredVariables),
+            (library) {
+              final unit = _parsedUnitFromUri(_session, library);
+              if (unit == null) return null;
+              final languageVersionToken = unit.languageVersionToken;
+              if (languageVersionToken == null) return null;
+              // dart:ui has no package name. So we cannot trivially look it up in
+              // the allowed experiments. We just assume it is opted in.
+              if (library == Uri.parse('dart:ui')) return null;
```

Comment:
Let's make this assumption for all `dart:` libraries, let's assume they are opted-in and have no langauge-version-markers.

Maybe we can say that if `!library.isScheme('package')` then we `return null`, that way we never have to deal `package = null` (I assume it's only null for `dart:` libraries).

Reply:
Good point
Done!